**Pandas**

Is a python library for working with data sets.

Can perform:
1. Analyzing
2. Cleaning
3. Visualization
4. Exploring and Manipulating

**Why Use Pandas**

*Allow us to analyze big data and make statistical conclusions

*Able to clean messy data and make them readable

*Relevant Data is important in Data Science


*Data Cleaning*- Deleting rows that are not relevant or contains wrong values i.e empty or NULL Values


In [1]:
#lets import
import pandas as pd

In [2]:
#lets create a simple data frame
#To create a data frame we use a dictionary
my_cars={
    "car":['BMW','Ferari','Mazda'],
    "prices":[190000,200000,10000]
}

df_cars=pd.DataFrame(my_cars)
print(df_cars)


      car  prices
0     BMW  190000
1  Ferari  200000
2   Mazda   10000


**Series**

A pandas series is like a column in a table

It is a one dimension array holding data of any type



In [6]:
#simple series
a=[1,2,3,4]

#add the column using pandas
df_a=pd.Series(a)

#display the series
print(df_a)

0    1
1    2
2    3
3    4
dtype: int64


**Labels**

If nothing else is specified the values are labled with their index numbers.

Index starts at '0'

This labels(index) Can be used to access a specified value.

*N/B*: This only works on a series not a dataframe. When working with dataframes use iloc[] instead

In [7]:
#lets access value "4" in our df_a
print(df_a[3])

4


In [9]:
#lets access information on "Ferari" in our df_cars------Dataframe
print(df_cars.iloc[2])

car       Mazda
prices    10000
Name: 2, dtype: object


**Create Labels**

With the index argument we can label our own labels

In [10]:
#create a series
a=[100,200,300]

#now lets add our label
myvar=pd.Series(a, index=['Juja','Thika','Nyeri'])

#display
print(myvar)

Juja     100
Thika    200
Nyeri    300
dtype: int64


In [11]:
#accessing items of the series above with user defined labeld
print(myvar['Thika'])

200


**Key/Value Objects as a series**

Using a dictionary when creating a series

*Note:* The keys of the dictionaries become the labels


In [13]:
#lets create a dictionary to pass to our series

calories={"day1":420,"day2":300,"day3":400, "day4":350,"day5":388}

#add the dictionary to a pandas series
df_calories=pd.Series(calories)

#display
print(df_calories)

day1    420
day2    300
day3    400
day4    350
day5    388
dtype: int64


In [14]:
#we can select only what we need to be in our series using the index
df_new_calories=pd.Series(calories, index=['day1','day3'])

#display
print(df_new_calories)

day1    420
day3    400
dtype: int64


**DataFrames**

Data sets in pandas are usually multi dimensional tables called DataFrames

*NOTE*: Series is like a column while a dataframe is the whole table

In [16]:
#Create a data frame from two series EG Students and there fee balance

student={
    "Name":["Allan","Jaba","Mark"],
    "Fee_Balance":[10000, 5799,11500]
}

#load the dataframe
df_students=pd.DataFrame(student, index=['BIT','BBC','BMIT'])

#display
print(df_students)

       Name  Fee_Balance
BIT   Allan        10000
BBC    Jaba         5799
BMIT   Mark        11500


**More on DataFrames...**

Is a two dimension data structure with rows and columns


In [17]:
#create dataframe with two series
data={
    "calories":[420,380,390],
    "duration(mins)":[50,40,45]
}

df_exercise=pd.DataFrame(data)

#display the data frame
print(df_exercise)

   calories  duration(mins)
0       420              50
1       380              40
2       390              45


**Locate Row**

Dataframe as shown from the results above is a table with rows and columns

Pandas use the **loc** attribute to return one or more specified row(s)

In [20]:
#refer to the row index
print(df_exercise.loc[1])

calories          380
duration(mins)     40
Name: 1, dtype: int64


In [22]:
#this example below wil return a pandas series(1 dimension remember)
print(df_exercise.loc[[0,2]])  #we are using a list of indexes

   calories  duration(mins)
0       420              50
2       390              45


In [26]:
#with the "index" argument we can label our indexes

company={
    "Emp_Name":['Allan','Jaba','Mark'],
    "Languages":['Python','JavaScript','Bash'],
    "Salary":[40000, 86999, 150000]
}

#load the dataframe above(Collection of series)
df_company=pd.DataFrame(company,index=['BIT','BBC','Cyber-Sec'])

#display the dataframe
print(df_company)

          Emp_Name   Languages  Salary
BIT          Allan      Python   40000
BBC           Jaba  JavaScript   86999
Cyber-Sec     Mark        Bash  150000


In [27]:
#Now lets use the loc attribute to get information on employee Jaba using the index 'BBC'
print(df_company.loc['BBC'])

Emp_Name           Jaba
Languages    JavaScript
Salary            86999
Name: BBC, dtype: object


In [28]:
#To save into a CSV file do. NOTE:FIle is saved inthe same directory as the Jupiter Notebook
df_company.to_csv('Dream_Company.csv', index=True)   #set the index to true to keep the default

**Load Files into A DataFrame**

If your data is stored into a file, Pandas supports a number of file extensions to load our data into a dataframe.

Lets manipulate the above 'Dream_Company.csv'

In [32]:
#read the csv file
df_employees=pd.read_csv('Dream_Company.csv', index_col=0)    #this index_col argument tells pandas to use the first column as our index

#dislay the loaded data
df_employees

,Emp_Name,Languages,Salary
BIT,Allan,Python,40000
BBC,Jaba,JavaScript,86999
Cyber-Sec,Mark,Bash,150000


In [33]:
#if you want to reset the column to be index values instead of your first column from loaded data becoming the index
df_employees_reset=df_employees.reset_index().rename(columns={'index':'Course'})

#display
print(df_employees_reset)

      Course Emp_Name   Languages  Salary
0        BIT    Allan      Python   40000
1        BBC     Jaba  JavaScript   86999
2  Cyber-Sec     Mark        Bash  150000


In [34]:
print(df_employees_reset.loc[0])

Course          BIT
Emp_Name      Allan
Languages    Python
Salary        40000
Name: 0, dtype: object


**Pandas Read CSV**

Simple way to store big data sets is to use a CSV file(Comma Separated Files)

**Tips**:to_string() prints the entire data frame




In [35]:
#load our dataframe
df_data=pd.read_csv('sample_data/data.csv')

#display
print(df_data.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

In [36]:
#display the first 5 rows. We use the head() method
print(df_data.head(5))

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0


In [37]:
#display the last 5 rows we use the 'tail()' method
print(df_data.tail(5))

     Duration  Pulse  Maxpulse  Calories
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4


**Read JSON**

Big datasets are often stored and extracted as json

*N/B*: JSON objects look similar to python dictionaries(has a key and value)

In [38]:
#load the json using pandas
df_json=pd.read_json('sample_data/data.js')

#display
print(df_json)

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.4
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


**Info About the Data**

info() method provides us with information about the dataframe

169 entries refers to the rows inthe table while Data Columns(4) is the number of columns in our table



In [45]:
df_json.info()   #we can identify null values in Calories column as there are some rows with missing values

<class 'pandas.core.frame.DataFrame'>
Index: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 6.6 KB


In [40]:
df_employees.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3 entries, BIT to Cyber-Sec
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Emp_Name   3 non-null      object
 1   Languages  3 non-null      object
 2   Salary     3 non-null      int64 
dtypes: int64(1), object(2)
memory usage: 204.0+ bytes


In [44]:
#The isnull() method can be used to get the number of null values in a certain column
df_json['Calories'].isnull().sum()   #missin 5 values


np.int64(5)

**Elements in a Data Frame**

When someone asks: “How many elements does this DataFrame have?”

Refers to the total number of data points(cells), not just rows
In Pandas:

Number of rows → df.shape[0]

Number of columns → df.shape[1]

Total elements (rows × columns) → df.size

In [19]:
import pandas as pd


company = {
    "Emp_Name": ['Allan', 'Jaba', 'Mark'],
    "Languages": ['Python', 'JavaScript', 'Bash'],
    "Salary": [40000, 86999, 150000]
}

#load the dataframe
df_company=pd.DataFrame(company)

#display
print(df_company)

  Emp_Name   Languages  Salary
0    Allan      Python   40000
1     Jaba  JavaScript   86999
2     Mark        Bash  150000


In [3]:
#now display the number of rows and columns
df_company.shape

(3, 3)

In [4]:
#display rows only
print(f'ROW: {df_company.shape[0]}')

ROW: 3


In [5]:
#display columns only
print(f'COLUMNS: {df_company.shape[1]}')

COLUMNS: 3


In [6]:
#get the total elements inthe dataframe
print(f'Total Elements inthe dataframe: {df_company.size}')

Total Elements inthe dataframe: 9


**Getting Max() and Min()**

Max() method is used to get the highest value in a column. Check to the column data type before using this

Min() method is used to identify the lowest value in a column

In [8]:
df_company['Salary'].max()  #gets the highest paid employee

150000

In [9]:
df_company['Salary'].min()   #gets the lowest paid employee

40000

In [14]:
# We can use sort_values method to sort the columns by the one with highest salary
df_company.sort_values(by='Salary', ascending=False)

,Emp_Name,Languages,Salary
2,Mark,Bash,150000
1,Jaba,JavaScript,86999
0,Allan,Python,40000


In [20]:
#Let set the first column in our above dataframe to be our index
#Specify the column name followed by the inplace attribute to make the changes permanent
df_company.set_index('Emp_Name', inplace=True)

In [21]:
#display
df_company

,Languages,Salary
Emp_Name,,
Allan,Python,40000
Jaba,JavaScript,86999
Mark,Bash,150000


In [28]:
df_company.loc['Mark'] #If we want now to access a value

,Mark
Languages,Bash
Salary,150000


In [30]:
#Lets use a condition check the which row has the salary == 40000
df_company[df_company['Salary']==40000]

,Languages,Salary
Emp_Name,,
Allan,Python,40000


In [32]:
#Here the first argument inside the loc is the index, followed by the column we want to check
df_company.loc['Mark', 'Salary']

np.int64(150000)

In [33]:

#who has a salary of 40000, lets assume we are iterating through a large column
names=['Mark','Jaba','Allan']

#write a for loop to loop through the column index and an if condition inside the for loop to check if someone is earning 40000
for name in names:
  #here we pass in the index and the row we want to check
  if df_company.loc[name, 'Salary']==40000:
    print(f'correct answer: {name}')
  else:
    print(f'{name}: not it')
  print()

Mark: not it

Jaba: not it

correct answer: Allan



In [36]:
#what is the highest possible length of words in our index
max_lenght=max(df_company.index.str.len())

print(max_lenght)

5


**ADDING A New Column**

we add a new column by adding the column name in df[] and values we are going to add into the data frame

E.G df['school']=['jkuat','uon','kibabi']

In [42]:
#lets try the above
df_company['school']=['jkuat','uon','kibabi']

#display
print(df_company)

           Languages  Salary  school
Emp_Name                            
Allan         Python   40000   jkuat
Jaba      JavaScript   86999     uon
Mark            Bash  150000  kibabi


In [49]:
#lets make it interesting  with a new column that has 10% of salary for SHA
df_company['SHA_Leavy']=(df_company['Salary']*2.75/100).round(2)    #this rounds to two decimal places

#display
print(df_company)

           Languages  Salary  school  SHA_Leavy
Emp_Name                                       
Allan         Python   40000   jkuat    1100.00
Jaba      JavaScript   86999     uon    2392.47
Mark            Bash  150000  kibabi    4125.00


**MODE**

In a column we may want to identify the most frequent value

We use the mode() method for this

In [37]:
data = {
    "Emp_Name": ["Allan", "Jaba", "Mark", "Allan"],
    "Salary": [40000, 86999, 150000, 40000]
}

df_startup = pd.DataFrame(data)

In [39]:
#check salary column
df_startup['Salary'].mode()

,Salary
0,40000


In [40]:
#check the employee column
df_startup['Emp_Name'].mode()

,Emp_Name
0,Allan
